### DISCUSSION TITLE
How to specify params path using nrunner?

### DISCUSSION THREAD

**Discussion by marioaag** [avocado-framework/avocado#4970] [Q&A] [answered]
I have being trying to get the parameters specifying a path without success and looking at the documentation it talks about the --mux-yaml option to make it available but it requires the mux plugin 

```

self.params.get('param_name', 'path_to_yaml_file')
```

Is it a way to specify the path where the yaml file whit the params values is located?
What is the default directory where abocado is going to look for the files with params?

**Comment by willianrampazzo** [comment]
Hi @marioaag, the path, in this case, is the path of the parameter in the params tree. I gave an example about the use of mux to another user here: https://github.com/avocado-framework/avocado/issues/4920#issuecomment-921149661 . If you are using a single parameter, `-p` command-line option is what you are looking for.

Another example, much more complex, is our test on supported images for the vmimage module: `selftests/pre_release/tests/vmimage.py` (https://github.com/avocado-framework/avocado/blob/master/selftests/pre_release/tests/vmimage.py), the yml file is at `selftests/pre_release/tests/vmimage.py.data/variants.yml` (https://github.com/avocado-framework/avocado/blob/master/selftests/pre_release/tests/vmimage.py.data/variants.yml).

**Reply by marioaag** [reply]
@willianrampazzo, thanks for the examples! 

What I'm looking for is be able to define default parameters values out of the 'test.py' file if them are not already provided in the process of the generation of the runner.

Some background:
I'm generating the runners and the suits through the API and I'm using nrunner

Base on this I have a another couple of question...
The default location for the yaml parameters file for an avocado instrumented tes is 'test_name.py.data/test_name.yaml'?
It will be used if don't have the -m switch?
Is the mux plugin already available for nrunner?

**Reply by willianrampazzo** [reply]
> @willianrampazzo, thanks for the examples!
> 
> What I'm looking for is be able to define default parameters values out of the 'test.py' file if them are not already provided in the process of the generation of the runner.
> 
> Some background:
> I'm generating the runners and the suits through the API and I'm using nrunner
> 
> Base on this I have a another couple of question...
> The default location for the yaml parameters file for an avocado instrumented tes is 'test_name.py.data/test_name.yaml'?
> It will be used if don't have the -m switch?

You can add the yml file to the test data folder, or, as you are using the Job API, you can set the folder manually. See https://github.com/avocado-framework/avocado/blob/master/examples/jobs/sleepjob_json_varianter.py for an example using a json file.

> Is the mux plugin already available for nrunner?

Yes, it is supported by nrunner. I added the support some weeks ago.

**Reply by marioaag** [reply]
So if I don't specify -m switch it should take the parameters form 'test_name.py.data/test_name.yaml' by default?

Using something like:
```
self.params.get('param_name', '*')
```

**Reply by willianrampazzo** [reply]
> So if I don't specify -m switch it should take the parameters form 'test_name.py.data/test_name.yaml' by default?
> 
> Using something like:
> 
> ```
> self.params.get('param_name', '*')
> ```

No, you still need to set the correct file to load. Following is a complete example.

**sleep.py**

```python
#!/usr/bin/env python3

import sys
import time

from avocado import Test
from avocado.core.job import Job

config = {'run.references': ['sleep_test.py:PassTest.test'],
          'yaml_to_mux.files': ['sleep_test.py.data/sleep_test.yml']}


with Job.from_config(job_config=config) as j:
    sys.exit(j.run())
```

**sleep_test.py**

```python
import time
from avocado import Test


class PassTest(Test):
    def test(self):
        length = float(self.params.get('length', default=3))
        time.sleep(length)
```

**sleep_test.py.data/sleep_test.yml**

```yml
!mux
small:
  length: 0.5
medium:
  length: 1.0
large:
  length: 2.0
```

When I run `sleep.py`, I get:

```
$ python sleep.py 
JOB ID     : fcbfb189f73efb397967526a1a51ff2eed0a7609
JOB LOG    : /home/wrampazz/avocado/job-results/job-2021-09-24T15.45-fcbfb18/job.log
 (1-1/3) sleep_test.py:PassTest.test;run-small-74d3: PASS (0.50 s)
 (1-2/3) sleep_test.py:PassTest.test;run-medium-7fac: PASS (1.00 s)
 (1-3/3) sleep_test.py:PassTest.test;run-large-8fc9: PASS (2.00 s)
RESULTS    : PASS 3 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /home/wrampazz/avocado/job-results/job-2021-09-24T15.45-fcbfb18/results.html
JOB TIME   : 4.02 s
```

**Reply by willianrampazzo** [reply]
I edited the previous comment with the correct information as the bug I mentioned was because of the spell difference in the YAML file.

**Reply by marioaag** [reply]
Thanks so much will give ti a try!!!

The yaml_to_mux.files needs to be specified at Job level or could be specified at runnable level?

**Reply by willianrampazzo** [reply, marked as answer]
> Thanks so much will give ti a try!!!
> 
> The yaml_to_mux.files needs to be specified at Job level or could be specified at runnable level?

You need to specify it on the Job level because it automatically creates new tasks.

**Comment by marioaag** [comment]
oooooo ic, it is not possible to use is as default value instead of hardcode the value in the TC.
